staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -191,40 +228,58 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 	return res, nil
 }
 
-func (a *Authenticator) tokenExpired() bool {
+func (a *Authenticator) credsExpired() bool {
 	if a.exp.IsZero() {
 		return false
 	}
 	return a.now().After(a.exp)
 }
 
 func (a *Authenticator) token() (string, error) {
+	creds, err := a.freshCreds()
+	if err != nil {
+		return "", err
+	}
+	return creds.token, nil
+}
+
+func (a *Authenticator) cert() (*tls.Certificate, error) {
+	creds, err := a.freshCreds()
+	if err != nil {
+		return nil, err
+	}
+	return creds.cert, nil
+}
+
+func (a *Authenticator) freshCreds() (*credentials, error) {
 	a.mu.Lock()
 	defer a.mu.Unlock()
-	if a.cachedToken != "" && !a.tokenExpired() {
-		return a.cachedToken, nil
+	if a.cachedCreds != nil && !a.credsExpired() {
+		return a.cachedCreds, nil
 	}
 
-	return a.getToken(nil)
+	if err := a.getCreds(nil); err != nil {
+		return nil, err
+	}
+	return a.cachedCreds, nil
 }
 
 // refresh executes the plugin to force a rotation of the token.
 func (a *Authenticator) refresh(token string, r *clientauthentication.Response) error {
 	a.mu.Lock()
 	defer a.mu.Unlock()
 
-	if token != a.cachedToken {
+	if token != a.cachedCreds.token {
```

Comment:
We need to document that 401s wont cause client certs to rotate. I guess we don't have the ability to revoke certs right now so it's not that big of a deal?

Reply:
Invalid client certs should fail at TLS handshake. There won't be an HTTP response.
And tls.Dial doesn't seem to provide a way to differentiate between "bad client cert" vs any regular network error?
Exec plugin could also do some cert caching, so re-execing it doesn't guarantee rotation.

I propose to leave this as is and maybe recommend short-lived certs?

Reply:
Ah of course. I don't think we should worry about this for now.

Reply:
i think it's possible to get a 401 if the connection was established before the client cert expired but the cert is now expired.

Reply:
Does the server check client cert after handshake at all?
If yes and this edge case gets hit too often, exec plugin can simply return expirationTime = certExpirationTime - 1d. That should rotate all connections before cert expiry.

Reply:
> Invalid client certs should fail at TLS handshake. There won't be an HTTP response.

that's not accurate... because the apiserver allows for both TLS and token auth, it does the client cert issuer/expiry/key-usage check post-handshake. only private key proof-of-possession checks are done during the handshake.